File: concepts/08_async_and_concurrency/01_async_await_and_result_propagation/python/test_02_task_ownership_and_failure_collection.py
"""任务所有权、结果复用与失败收集。

共同问题：异步结果能否被多个观察者复用；组合失败是否取消其他工作；
调用方如何选择快速失败或把失败作为结果收集。
"""

# polyglot-family: async_and_concurrency
# polyglot-concept: async_await_and_result_propagation
# polyglot-related: languages/python/stdlib/095-103_networking_and_interprocess_communication/
# polyglot-related+: test_095_asyncio_coroutines_tasks_futures_threads_and_errors.py

import asyncio


def test_task_result_can_be_awaited_repeatedly_after_completion():
    async def scenario():
        calls = 0

        async def work():
            nonlocal calls
            calls += 1
            return 42

        task = asyncio.create_task(work())
        first = await task
        second = await task
        return first, second, calls

    assert asyncio.run(scenario()) == (42, 42, 1)

    # coroutine object 本身完成后不可复用；Task/Future 持有完成状态，可供多个观察者读取。


def test_gather_can_collect_failures_in_input_order():
    async def scenario():
        async def fail():
            raise ValueError("failed")

        return await asyncio.gather(
            asyncio.sleep(0, result=1),
            fail(),
            return_exceptions=True,
        )

    results = asyncio.run(scenario())

    assert results[0] == 1
    assert isinstance(results[1], ValueError)
    assert str(results[1]) == "failed"

    # return_exceptions=False 会向等待者传播首个失败，但不等于结构化取消所有其他任务；
    # 所有权与取消策略必须由调用方显式安排。

File: concepts/08_async_and_concurrency/01_async_await_and_result_propagation/rust/test_02_task_ownership_and_failure_collection.rs
// polyglot-family: async_and_concurrency
// polyglot-concept: async_await_and_result_propagation
// polyglot-related: languages/rust/tests/language/12_async_future_pin_and_cancellation/
// polyglot-related+: test_094_result_propagation_and_joining_futures.rs
//
// 共同问题：多个任务由谁等待；失败是否被遗漏；结果收集何时结束。
// 对照观察：std 没有 task tree；owner must retain/poll futures or join threads and collect each Result explicitly.

#[test]
fn comparison() {
    let handles: Vec<_> = [Ok::<_, &'static str>(1), Err("failed"), Ok(3)]
        .into_iter()
        .map(|result| std::thread::spawn(move || result))
        .collect();
    let results: Vec<_> = handles
        .into_iter()
        .map(|handle| handle.join().unwrap())
        .collect();
    assert_eq!(results, [Ok(1), Err("failed"), Ok(3)]);
    assert_eq!(results.iter().filter(|result| result.is_err()).count(), 1);
}
